src/main/java/com/salesforce/functions/jvm/sdk/bulk/BulkApi.java

```diff
@@ -0,0 +1,141 @@
+   */
+  @SuppressWarnings("unused")
+  String createJob(Job job) throws BulkApiException;
+
+  /**
+   * Retrieves detailed information about a job.
+   *
+   * @param jobId The ID of the job to get information about.
+   * @return Detailed information about a job.
+   * @throws BulkApiException If an error occurred while querying job info.
+   */
+  @SuppressWarnings("unused")
+  JobInfo getJobInfo(String jobId) throws BulkApiException;
+
+  /**
+   * Retrieves a list of successfully processed records for a completed job.
+   *
+   * <p>The returned file is a CSV file that all the records that encountered an error while being
+   * processed by the job. Each row corresponds to a successfully processed record.
+   *
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_successful_results.htm">Bulk
+   *     API 2.0 and Bulk API Developer Guide</a>
+   * @param jobId The ID of the job to get successful record results for.
+   * @return A CSV file with successful records.
+   * @throws BulkApiException If an error occurred while querying successful record results.
+   */
+  @SuppressWarnings("unused")
+  File getJobSuccessfulRecordResults(String jobId) throws BulkApiException;
+
+  /**
+   * Retrieves a list of failed records for a completed job.
+   *
+   * <p>The returned file is a CSV file that all the records that encountered an error while being
+   * processed by the job. Each row corresponds to a failed record.
+   *
+   * @see <a
+   *     href="https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/get_job_failed_results.htm">Bulk
+   *     API 2.0 and Bulk API Developer Guide</a>
+   * @param jobId The ID of the job to get failed record results for.
+   * @return A CSV file with all failed records.
+   * @throws BulkApiException If an error occurred while querying failed record results.
+   */
+  @SuppressWarnings("unused")
+  File getJobFailedRecordResults(String jobId) throws BulkApiException;
+
+  /**
+   * Uploads data for a job using CSV data you provide.
+   *
+   * @param jobId The ID for the job to upload data for.
+   * @param file The CSV file to upload to the job.
+   * @throws BulkApiException If an error occurred uploading the data.
+   * @throws IOException If an error occurred reading data from the provided file.
+   */
+  @SuppressWarnings("unused")
+  void uploadJobData(String jobId, File file) throws BulkApiException, IOException;
+
+  /**
+   * Uploads data for a job using CSV data you provide.
+   *
+   * @param jobId The ID for the job to upload data for.
+   * @param inputStream The CSV data as a stream to upload to the job.
+   * @throws BulkApiException If an error occurred uploading the data.
+   * @throws IOException If an error occurred reading data from the provided stream.
+   */
+  @SuppressWarnings("unused")
+  void uploadJobData(String jobId, InputStream inputStream) throws BulkApiException, IOException;
```

Comment:
I think I prefer this signature to the `File` version. This seems more flexible, as developers could turn a file into a stream pretty easily (or at least I think it's easy in Java /shrug). Do we need to support both signatures, or could we get away with just one?

Reply:
We could get away with just one - implementation wise, all of them would probably use the `InputStream` version anyway. For Java, having a direct `File` version seems nice to have and isn't hard to implement.

Reply:
I removed the non-InputStream variants in [f39afd8](https://github.com/forcedotcom/sf-fx-sdk-java/pull/64/commits/f39afd874afe5e9cd72a8ddb3cb738564ff8adef).